File: c++/remove_tree_nodes.cpp
/**
* Created on 7/23/16.
* Given a binary tree and a value list,
* return the tree root values if the nodes
* of the values in the value list are deleted.
*
* For example, tree
*      1
*     / \
*    2  5
*   / \  \
*  3  4  6
*
*  list = [2, 5],
*  return
*  [1, 3, 4, 6]
*/

#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

struct TreeNode {
    TreeNode(int val) {
        this->val = val;
        left = nullptr;
        right = nullptr;
    }
    TreeNode* left;
    TreeNode* right;
    int val;
};

class Solution {
public:
    vector<TreeNode*> removeTreeNodes(TreeNode* root, const vector<int>& nodeVals) {
        unordered_set<int> nodeValSet;
        for (int v : nodeVals) {
            nodeValSet.insert(v);
        }
        vector<TreeNode*> res;
        root = removeTreeNodesHelper(root, nodeValSet, res);
        if (root != nullptr) {
            res.push_back(root);
        }
        return res;
    }

private:
    // Post-order traversal of removing tree nodes
    TreeNode* removeTreeNodesHelper(TreeNode* node, const unordered_set<int>& nodeValSet, vector<TreeNode*>& res) {
        if (nullptr == node) return nullptr;
        node->left = removeTreeNodesHelper(node->left, nodeValSet, res);
        node->right = removeTreeNodesHelper(node->right, nodeValSet, res);
        if (nodeValSet.count(node->val)) {
            if (node->left != nullptr) res.push_back(node->left);
            if (node->right != nullptr) res.push_back(node->right);
            delete node;
            node = nullptr;
        }
        return node;
    }
};

int main()
{
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(5);
    root->left->left = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->right->right = new TreeNode(6);
    vector<int> values = {2, 5};
    Solution sol;
    vector<TreeNode*> res = sol.removeTreeNodes(root, values);
    for (TreeNode* node : res) {
        cout << node->val << ' ';
    }
    cout << endl;
}

File: java/RemoveTreeNodes.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created on 7/23/16.
 * Given a binary tree and a value list,
 * return the tree root values if the nodes
 * of the values in the value list are deleted.
 *
 * For example, tree
 *      1
 *     / \
 *    2  5
 *   / \  \
 *  3  4  6
 *
 *  list = [2, 5],
 *  return
 *  [1, 3, 4, 6]
 */
public class RemoveTreeNodes {

    public static void main(String[] args) {
        int[] list = {2, 5, 1};
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(5);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.right = new TreeNode(6);
        RemoveTreeNodes app = new RemoveTreeNodes();
        System.out.println(app.removeTreeNodes(root, list).toString());
    }

    public List<Integer> removeTreeNodes(TreeNode root, int[] nums) {
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) {
            numSet.add(num);
        }
        List<Integer> res = new ArrayList<>();
        root = removeTreeNodesHelper(root, numSet, res);
        if (root != null) res.add(root.val);
        return res;
    }

    /**
     * Post order traversal to remove
     * @param node
     * @param numSet
     * @param res
     * @return
     */
    private TreeNode removeTreeNodesHelper(TreeNode node, Set<Integer> numSet, List<Integer> res) {
        if (null == node) return null;
        node.left = removeTreeNodesHelper(node.left, numSet, res);
        node.right = removeTreeNodesHelper(node.right, numSet, res);
        if (numSet.contains(node.val)) {
            if (node.left != null) res.add(node.left.val);
            if (node.right != null) res.add(node.right.val);
            return null;
        }
        return node;
    }
}
